primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,859 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Trie Cache
+//!
+//! Provides an implementation of the [`TrieCache`](trie_db::TrieCache) trait.
+//! The implementation is split into three types [`SharedTrieCache`], [`LocalTrieCache`] and
+//! [`TrieCache`]. The [`SharedTrieCache`] is the instance that should be kept around for the entire
+//! lifetime of the node. It will store all cached trie nodes and values on a global level. Then
+//! there is the [`LocalTrieCache`] that should be kept around per state instance requested from the
+//! backend. As there are very likely multiple accesses to the state per instance, this
+//! [`LocalTrieCache`] is used to cache the nodes and the values before they are merged back to the
+//! shared instance. Last but not least there is the [`TrieCache`] that is being used per access to
+//! the state. It will use the [`SharedTrieCache`] and the [`LocalTrieCache`] to fulfill cache
+//! requests. If both of them don't provide the requested data it will be inserted into the
+//! [`LocalTrieCache`] and then later into the [`SharedTrieCache`].
+//!
+//! The [`SharedTrieCache`] is bound to some maximum number of bytes. It is ensured that it never
+//! runs above this limit. However as long as data is cached inside a [`LocalTrieCache`] it isn't
+//! taken into account when limiting the [`SharedTrieCache`]. This means that for the lifetime of a
+//! [`LocalTrieCache`] the actual memory usage could be above the allowed maximum.
+
+use crate::{Error, NodeCodec};
+use hash_db::Hasher;
+use lru::LruCache;
+use nohash_hasher::{BuildNoHashHasher, IntMap, IntSet};
+use parking_lot::{Mutex, MutexGuard, RwLock, RwLockReadGuard};
+use std::{
+	collections::{hash_map::Entry, HashMap, HashSet},
+	hash::{BuildHasher, Hasher as _},
+	mem,
+	sync::Arc,
+};
+use trie_db::{node::NodeOwned, CachedValue};
+
+const LOG_TARGET: &str = "trie-cache";
+
+lazy_static::lazy_static! {
+	static ref RANDOM_STATE: ahash::RandomState = ahash::RandomState::default();
+}
+
+/// No hashing [`lru::LruCache`].
+///
+/// The key is an `u64` that is expected to be unique.
+type NoHashingLruCache<T> = lru::LruCache<u64, T, BuildNoHashHasher<u64>>;
```

Comment:
We could use `BuildNoHashHasher` with hashmap, hashset instead of depending on `nohash_hasher`.

Reply:
I mean between using `nohash_hasher` and just using a hash map with direct https://github.com/paritytech/parity-db/blob/b300cbbef3b1c8378601014c7236a83f3d0882b9/src/lib.rs#L40 there is no difference, but using identity hasher seems more straightforward to me. (it is just a nitpick, but with the additional deps, that is one more source code to check)

Reply:
We use nohash-hasher anyway already in our deps.